### ISSUE TITLE
Create Java Client SDK 

### ISSUE THREAD

**Issue by tanertopal** [adap/flower#342]
Create a Java Client SDK to make it easier to start with Java/Android.

**Comment by lucastliu** [comment]
Any plans to implement this? Alternatively, is there anywhere I can find examples / code of Android usage of Flower?

**Comment by danieljanes** [comment]
Yes, there will be progress on this in the coming weeks, but we don't have a fixed target date yet.

The general approach to using any language with Flower is:
1. Compile `transport.proto` for the language of choice (using the ProtoBuf/gRPC compiler)
2. Use the generated code to establish a connection to the Flower server
3. Handle the incoming messages (mainly `fit` and `evaluate`) and return the results to the server

@lucastliu which ML framework are you planning to use with your Java clients?

**Comment by lucastliu** [comment]
Thanks for the update. I am trying to get an idea of where the flower repository stands (since the publishing of the paper "Flower: A Friendly Federated Learning Research Framework"). There was mention of a "Java SDK (Android)", and I was wondering if this was it. I understand that I can fulfill the raw interface with almost any strategy implementations, but I want to understand if perhaps certain devices / languages had more support / "out-of-box" code. 

On a related note, would you consider implementing flower to be equally easy on Android vs. iOS as of today? If not, which parts of flower make one easier as of now?

**Comment by danieljanes** [comment]
You're right that the event protocol can be implemented in almost any language. Flower currently uses bi-directional gRPC streaming for client/server communication, so languages/platforms with good gRPC support have an advantage. For other cases, there's still the option to add an additional communication stack, but that would be a bit more effort.

Regarding your second question, there's no inherent difference when it comes to supporting Android or iOS - both require you to compile the `.proto` definitions to the target language (Java/Kotlin vs Objective-C/Swift). Once the gRPC/ProtoBuf definitions are compiled, one just needs to establish a connection to a Flower server and handle the incoming messages (which can be seen in the `.proto` definition).

By the way, we're open to merge contributions and can support you in case you'd want to tackle this.

**Comment by WilliamLindskog** [comment]
@danieljanes does current https://github.com/adap/flower/tree/main/examples/android cover this?

**Comment by danieljanes** [comment]
@WilliamLindskog yes, but the example is not compatible with the latest version of Flower. Let's keep this issue open until the Kotlin SDK out of preview.

**Comment by WilliamLindskog** [comment]
@danieljanes ok 👍🏽 Adding to Future Release, thus moving to `Planned`.
